File: backend/src/modules/subscriber/subscriber.service.ts
import { Injectable, NotFoundException, ConflictException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, Like } from 'typeorm';
import { createHash } from 'crypto';
import { Subscriber } from './entities/subscriber.entity';
import { CreateSubscriberDto } from './dto/create-subscriber.dto';
import { UpdateSubscriberDto } from './dto/update-subscriber.dto';
import { PaginationDto, PaginatedResult } from '../../common/dto/pagination.dto';

@Injectable()
export class SubscriberService {
  constructor(
    @InjectRepository(Subscriber)
    private subscriberRepository: Repository<Subscriber>,
  ) {}

  /**
   * 生成 HA1 哈希
   * HA1 = MD5(username:realm:password)
   */
  private generateHa1(username: string, domain: string, password: string): string {
    return createHash('md5')
      .update(`${username}:${domain}:${password}`)
      .digest('hex');
  }

  /**
   * 生成 HA1B 哈希
   * HA1B = MD5(username@realm:realm:password)
   */
  private generateHa1b(username: string, domain: string, password: string): string {
    return createHash('md5')
      .update(`${username}@${domain}:${domain}:${password}`)
      .digest('hex');
  }

  /**
   * 创建订阅者
   */
  async create(createDto: CreateSubscriberDto): Promise<Subscriber> {
    // 检查是否已存在
    const existing = await this.subscriberRepository.findOne({
      where: { username: createDto.username, domain: createDto.domain },
    });

    if (existing) {
      throw new ConflictException('用户已存在');
    }

    const subscriber = this.subscriberRepository.create({
      username: createDto.username,
      domain: createDto.domain,
      password: createDto.password,
      ha1: this.generateHa1(createDto.username, createDto.domain, createDto.password),
      ha1b: this.generateHa1b(createDto.username, createDto.domain, createDto.password),
      emailAddress: createDto.emailAddress,
      rpid: createDto.rpid,
    });

    return this.subscriberRepository.save(subscriber);
  }

  /**
   * 分页查询订阅者
   */
  async findAll(paginationDto: PaginationDto): Promise<PaginatedResult<Subscriber>> {
    const { page, limit, keyword } = paginationDto;
    const skip = (page - 1) * limit;

    const where: any = {};
    if (keyword) {
      where.username = Like(`%${keyword}%`);
    }

    const [items, total] = await this.subscriberRepository.findAndCount({
      where,
      skip,
      take: limit,
      order: { id: 'DESC' },
    });

    return new PaginatedResult(items, total, page, limit);
  }

  /**
   * 根据 ID 查询订阅者
   */
  async findOne(id: number): Promise<Subscriber> {
    const subscriber = await this.subscriberRepository.findOne({ where: { id } });
    if (!subscriber) {
      throw new NotFoundException('用户不存在');
    }
    return subscriber;
  }

  /**
   * 根据用户名和域查询
   */
  async findByUsernameAndDomain(username: string, domain: string): Promise<Subscriber> {
    const subscriber = await this.subscriberRepository.findOne({
      where: { username, domain },
    });
    if (!subscriber) {
      throw new NotFoundException('用户不存在');
    }
    return subscriber;
  }

  /**
   * 更新订阅者
   */
  async update(id: number, updateDto: UpdateSubscriberDto): Promise<Subscriber> {
    const subscriber = await this.findOne(id);

    if (updateDto.password) {
      subscriber.password = updateDto.password;
      subscriber.ha1 = this.generateHa1(subscriber.username, subscriber.domain, updateDto.password);
      subscriber.ha1b = this.generateHa1b(subscriber.username, subscriber.domain, updateDto.password);
    }

    if (updateDto.emailAddress !== undefined) {
      subscriber.emailAddress = updateDto.emailAddress;
    }

    if (updateDto.rpid !== undefined) {
      subscriber.rpid = updateDto.rpid;
    }

    return this.subscriberRepository.save(subscriber);
  }

  /**
   * 删除订阅者
   */
  async remove(id: number): Promise<void> {
    const subscriber = await this.findOne(id);
    await this.subscriberRepository.remove(subscriber);
  }

  /**
   * 批量删除订阅者
   */
  async batchRemove(ids: number[]): Promise<void> {
    await this.subscriberRepository.delete(ids);
  }

  /**
   * 获取统计信息
   */
  async getStats(): Promise<{ total: number; domains: { domain: string; count: number }[] }> {
    const total = await this.subscriberRepository.count();
    
    const domains = await this.subscriberRepository
      .createQueryBuilder('s')
      .select('s.domain', 'domain')
      .addSelect('COUNT(*)', 'count')
      .groupBy('s.domain')
      .getRawMany();

    return { total, domains };
  }
}

File: kam-admin-module-user/src/main/java/io/kamailio/admin/modules/subscriber/service/SubscriberService.java
package io.kamailio.admin.modules.subscriber.service;

import io.kamailio.admin.common.exception.BusinessException;
import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.common.dto.PaginationDto;
import io.kamailio.admin.modules.subscriber.dto.CreateSubscriberDto;
import io.kamailio.admin.modules.subscriber.dto.UpdateSubscriberDto;
import io.kamailio.admin.modules.subscriber.entity.Subscriber;
import io.kamailio.admin.modules.subscriber.repository.SubscriberRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.List;
import java.util.Map;

@Service
public class SubscriberService {

    private final SubscriberRepository repository;

    public SubscriberService(SubscriberRepository repository) {
        this.repository = repository;
    }

    private static String md5Hex(String input) {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] digest = md.digest(input.getBytes(java.nio.charset.StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder();
            for (byte b : digest) sb.append(String.format("%02x", b));
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    private String generateHa1(String username, String domain, String password) {
        return md5Hex(username + ":" + domain + ":" + password);
    }

    private String generateHa1b(String username, String domain, String password) {
        return md5Hex(username + "@" + domain + ":" + domain + ":" + password);
    }

    @Transactional
    public Subscriber create(CreateSubscriberDto dto) {
        if (repository.findByUsernameAndDomain(dto.getUsername(), dto.getDomain()).isPresent()) {
            throw new BusinessException("用户已存在");
        }
        var s = new Subscriber();
        s.setUsername(dto.getUsername());
        s.setDomain(dto.getDomain());
        s.setPassword(dto.getPassword());
        s.setHa1(generateHa1(dto.getUsername(), dto.getDomain(), dto.getPassword()));
        s.setHa1b(generateHa1b(dto.getUsername(), dto.getDomain(), dto.getPassword()));
        s.setEmailAddress(dto.getEmailAddress());
        s.setRpid(dto.getRpid());
        return repository.save(s);
    }

    public PaginatedResult<Subscriber> findAll(PaginationDto dto) {
        Page<Subscriber> page;
        if (dto.getKeyword() != null && !dto.getKeyword().isBlank()) {
            page = repository.findByUsernameContaining(dto.getKeyword(),
                    PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by(Sort.Direction.DESC, "id")));
        } else {
            page = repository.findAll(PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by(Sort.Direction.DESC, "id")));
        }
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }

    public Subscriber findOne(Integer id) {
        return repository.findById(id).orElseThrow(() -> new BusinessException("用户不存在"));
    }

    public Subscriber findByUsernameAndDomain(String username, String domain) {
        return repository.findByUsernameAndDomain(username, domain).orElseThrow(() -> new BusinessException("用户不存在"));
    }

    @Transactional
    public Subscriber update(Integer id, UpdateSubscriberDto dto) {
        var s = findOne(id);
        if (dto.getPassword() != null) {
            s.setPassword(dto.getPassword());
            s.setHa1(generateHa1(s.getUsername(), s.getDomain(), dto.getPassword()));
            s.setHa1b(generateHa1b(s.getUsername(), s.getDomain(), dto.getPassword()));
        }
        if (dto.getEmailAddress() != null) s.setEmailAddress(dto.getEmailAddress());
        if (dto.getRpid() != null) s.setRpid(dto.getRpid());
        return repository.save(s);
    }

    @Transactional
    public void remove(Integer id) {
        repository.delete(findOne(id));
    }

    @Transactional
    public void batchRemove(List<Integer> ids) {
        repository.deleteAllById(ids);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> getStats() {
        long total = repository.count();
        var rows = repository.countByDomain();
        var domains = rows.stream().map(r -> Map.<String, Object>of("domain", r[0], "count", r[1])).toList();
        return Map.of("total", total, "domains", domains);
    }
}
